### PR TITLE
JENKINS-64418: add simple exponential backoff to the sleep for bitbucket api rate limits

### PR DESCRIPTION
**Pull request by b-dean**
When we have hit the rate limits for Bitbucket API it slows everything down to a crawl. Also other API requests (such as getting the Jenkinsfile for multibranch pipelines) are rate limited and slower.

Adding exponential backoff with a minimum of 100 ms and a max of 1 minute, should alleviate the constant retry loop hitting the limit every 5s.

See [JENKINS-64418](https://issues.jenkins.io/browse/JENKINS-64418).

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

**Comment by artheus** [comment]
Looking at Atlassians documentation for "Adjusting your code for rate limiting" I would argue that, of the three alternative ways proposed there, the 3rd option makes the most sense here.

See: https://confluence.atlassian.com/adminjiraserver/adjusting-your-code-for-rate-limiting-987143384.html

The 3rd option states that one can adjust the requests sent based on the request limit http headers.

> * `x-ratelimit-interval-seconds`: The time interval in seconds. You get a batch of new tokens every time interval.
> * `x-ratelimit-fillrate`: The number of tokens you get every time interval.
> * `retry-after`: The number of seconds you need to wait for new tokens. Make sure that your rate assumes waiting longer than this value.

I argue that this would be the way to go, due to the two reasons mentioned in the documentation about exponential backoff

> :no_entry:  High impact on a Jira instance because of concurrency. We’re assuming most active users will send requests whenever they’re available. This window will be similar for all users, making spikes in Jira performance. The same applies to threads — most will either be busy at the same time or idle.
>
> :no_entry:  Unpredictable. If you need to make a few critical requests, you can’t be sure all of them will be successful.

**Comment by bitwiseman** [comment]
@artheus 
Agreed, that is a better way. 
If you take a look at https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java it uses similar header info from GitHub to reason over rate limits and choose a strategy. However, that will take considerably more work.  Would you be interested in submitting a PR? 

Also, take a look at https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/187#issuecomment-741929228.  There might be a way to mitigate the behavior in some cases by using caching.

**Comment by artheus** [comment]
@bitwiseman I will start working on a PR for this.
I do not really agree to the use of caching as a part of the plugin though. I would much rather suggest/recommend users to use a caching reverse-proxy between Jenkins and Bitbucket, this way users would be in much better control of the caching layer, and the plugin would work as expected.
Creating a caching layer here would make this a much larger job, as we would have to make a large amount of separate decisions, e.g:
- In which scope to cache (per job, run, bitbucket project or globally)
- In RAM or on disk?
  - if on disk, where would be the best place to store caches?
- How to invalidate caches?
- What is a good timeout for caching data from which changes are expected?
- If a synchronization job is run, and cached data is used, that would be a complete waste of resources (RAM, CPU...)

I think that the scope for this issue should be just to add a way to properly comply with the Bitbucket rate limits.
We could create a separate issue, PR or project for a caching layer.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 1]
```diff
@@ -866,7 +869,8 @@ protected CloseableHttpResponse executeMethod(HttpHost host, HttpRequestBase htt
         httpMethod.setConfig(requestConfig.build());
 
         CloseableHttpResponse response = client.execute(host, httpMethod, requestContext);
-        while (response.getStatusLine().getStatusCode() == API_RATE_LIMIT_CODE) {
+        Random random = new Random();
+        for (int attempts = 0; response.getStatusLine().getStatusCode() == API_RATE_LIMIT_CODE; attempts++) {
```
Comment (bitwiseman) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:873]: The minimum sleep should remain 5 seconds.  Once the limit is hit, the client needs to a wait at least on the order of seconds before trying again.  

The maximum around 2 or 3 minutes sounds good.   

We don't currently limit the number of attempts.  That shouldn't change.  We should hit the max and stay there.

**Review by bitwiseman — CHANGES_REQUESTED** [review]
We need a test that exercises this.

**Comment by b-dean** [comment]
@artheus, that article  you linked to from Atlassian is for JIRA, not Bitbucket. I logged all the headers that come back from the request that has the 429 response and there's nothing that says `retry-after`

```
Server: nginx
Vary: Authorization, cookie, user-context
Cache-Control: no-cache, no-store
Content-Type: text/plain
X-B3-TraceId: 1a4e3ec6e8f2fe21
X-Dc-Location: ash2
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Date: Thu, 04 Feb 2021 21:55:33 GMT
X-Served-By: app-3002
X-Static-Version: 6d8cc6ef88d5
X-Credential-Type: apikey
X-Render-Time: 0.0227448940277
X-Accepted-OAuth-Scopes: repository
Connection: Keep-Alive
X-Version: 6d8cc6ef88d5
X-Request-Count: 152
X-Frame-Options: SAMEORIGIN
Content-Length: 46
```

I thought about `X-Request-Count` but [someone else asked Atlassian and they said it means nothing](https://community.developer.atlassian.com/t/what-does-x-request-count-means-in-bitbucket-api-response-header/34589).

**Comment by nfalco79** [comment]
Superseded by #927

**Pull request closed without merging**